hazelcast/src/main/java/com/hazelcast/util/collection/OAHashSet.java

```diff
@@ -0,0 +1,463 @@
+ * This {@link Set} implementation does not permit concurrent modifications
+ * during iteration.
+ *
+ * @param <E> The type of the elements stored in the set
+ */
+public class OAHashSet<E> extends AbstractSet<E> implements Set<E> {
+    private static final int DEFAULT_INITIAL_CAPACITY = 16;
+    private static final float DEFAULT_LOAD_FACTOR = 0.6F;
+
+    private final float loadFactor;
+
+    private int[] hashes;
+    private Object[] table;
+    private int resizeThreshold;
+    private int capacity;
+    private int mask;
+    private int size;
+    /**
+     * The version of this set. Used to detect concurrent modification when
+     * iterating over the elements in the set with {@link ElementIterator}
+     */
+    private int version;
+
+    /**
+     * Constructs an {@link OAHashSet} instance with default initial
+     * capacity and default load factor
+     *
+     * @see #DEFAULT_INITIAL_CAPACITY
+     * @see #DEFAULT_LOAD_FACTOR
+     */
+    public OAHashSet() {
+        this(DEFAULT_INITIAL_CAPACITY, DEFAULT_LOAD_FACTOR);
+    }
+
+    /**
+     * Constructs an {@link OAHashSet} instance with the specified
+     * initial capacity and with the default load factor
+     *
+     * @param initialCapacity the initial capacity of the set to be created
+     * @see #DEFAULT_LOAD_FACTOR
+     */
+    public OAHashSet(int initialCapacity) {
+        this(initialCapacity, DEFAULT_LOAD_FACTOR);
+    }
+
+    /**
+     * Constructs an {@link OAHashSet} instance with the specified
+     * initial capacity and load factor
+     *
+     * @param initialCapacity the initial capacity of the set to be created
+     * @param loadFactor      the load factor of the set to be created
+     */
+    public OAHashSet(int initialCapacity, float loadFactor) {
+        checkNotNegative(initialCapacity, "Illegal initial capacity: " + initialCapacity);
+        checkFalse(loadFactor <= 0 || Float.isNaN(loadFactor), "Illegal load factor: " + loadFactor);
+
+        this.capacity = QuickMath.nextPowerOfTwo(initialCapacity);
+        this.loadFactor = loadFactor;
+        this.resizeThreshold = (int) (capacity * loadFactor);
+        this.mask = capacity - 1;
+        this.hashes = new int[capacity];
+        this.table = new Object[capacity];
+    }
+
+    @Override
+    public boolean add(E element) {
+        return add(element.hashCode(), element);
+    }
+
+    /**
+     * Adds the specified element to this set if it is not already present.
+     * <p>
+     * This variant of {@link #add(Object)} acts as an optimisation to
+     * enable avoiding {@link #hashCode()} calls if the hash is already
+     * known on the caller side.
+     *
+     * @param hash         the hash of the element to be removed
+     * @param elementToAdd object to be removed from this set, if present
+     * @return <tt>true</tt> if this set did not already contain the specified
+     * element
+     * @see #add(Object)
+     */
+    public boolean add(int hash, E elementToAdd) {
```

Comment:
`add(element, hash)` seems like a more natural argument order, putting the main thing first.

Reply:
Changed the order of the arguments.